l2-contracts/contracts/ValidatorRegistry.sol

```diff
@@ -0,0 +1,177 @@
+            "Validator's inactivity delay has not passed"
+        );
+
+        // Remove from mapping.
+        delete validators[idx];
+
+        // Remove from array by swapping the last element (gas-efficient, not preserving order).
+        uint256 offset = idxOffset(idx);
+        validatorIndexes[offset] = validatorIndexes[validatorIndexes.length - 1];
+        validatorIndexes.pop();
+    }
+
+    // Inactivates a validator.
+    function inactivate(uint256 idx) public onlyOwner {
+        verifyExists(idx);
+        validators[idx].isInactive = true;
+        validators[idx].inactiveSince = epoch;
+    }
+
+    // Activates a validator.
+    function activate(uint256 idx) public onlyOwner {
+        verifyExists(idx);
+        validators[idx].isInactive = false;
+    }
+
+    // Changes the weight.
+    function changeWeight(uint256 idx, uint256 weight) public onlyOwner {
+        verifyExists(idx);
+        validators[idx].weight = weight;
+    }
+
+    // Changes the public key and PoP.
+    function changePublicKey(
+        uint256 idx,
+        bytes calldata pubKey,
+        bytes calldata pop
+    ) public onlyOwner {
+        verifyExists(idx);
+        verifyPoP(pubKey, pop);
+        validators[idx].pubKey = pubKey;
+        validators[idx].pop = pop;
+    }
+
+    // Creates a new committee list which will become the next committee list.
+    function setNextCommittee() external onlyOwner {
+        epoch += 1;
+
+        // Replace current committee with next committee
+        delete committee;
+        for (uint256 i = 0; i < nextCommittee.length; i++) {
+            committee.push(nextCommittee[i]);
+        }
+
+        // Populate `nextCommittee` based on active validators
+        delete nextCommittee;
+        for (uint256 i = 0; i < validatorIndexes.length; i++) {
+            uint256 idx = validatorIndexes[i];
+            Validator memory validator = validators[idx];
+            if (!validator.isInactive) {
+                nextCommittee.push(CommitteeValidator(idx, validator.weight, validator.pubKey));
+            }
+        }
+    }
+
+    // Finds the index offset of a validator in the `validatorIndexes` array.
+    function idxOffset(uint256 idx) private view returns (uint256) {
+        for (uint256 i = 0; i < validatorIndexes.length; i++) {
+            if (validatorIndexes[i] == idx) {
+                return i;
+            }
+        }
+        revert("Validator idx not found");
+    }
+
+    // Verifies that a validator exists.
+    function verifyExists(uint256 idx) private view {
+        require(validators[idx].pubKey.length != 0, "Validator doesn't exist");
+    }
+
+    function compareBytes(bytes storage a, bytes calldata b) private pure returns (bool) {
+        return keccak256(a) == keccak256(b);
+    }
+
+    // Verifies the proof-of-possession.
+    function verifyPoP(bytes calldata pubKey, bytes calldata pop) internal pure {
+        // TODO: implement or remove
```

Comment:
I didn't understand it first, but this requires https://eips.ethereum.org/EIPS/eip-2537 right? And IIUC the idea was that if they can't be validated in Solidity then the node will retrieve the data and filter out the invalids, although if they are all controlled by the owner, then I'm not sure why they would send an invalid key.

Reply:
Since it's an L2 contract, I don't think it's bound to EIP-2537, but yes, the needed precompiles are currently missing.

And yes, verification should be done off-chain AFAIU.

Key rotation can be done also by the node owner, and not just by the registries owner. Perhaps that's the answer to your question.

Reply:
That's correct.

Reply:
Okay, so if verification is done off-chain then we don't have to either implement nor remove it, that's why I asked. Sorry, I wasn't clear. 

Thank you for clarifying that we can have BLS earlier than L1 as well, good context 🙏

Reply:
Sorry, my misunderstanding came from whether you want to remove the PoP, or just its verification 😵